### PR TITLE
add print basic map functionality

### PR DESCRIPTION
**Pull request by ann26**
fix https://github.com/kartoza/LEDET_BIMS/issues/303 and fix https://github.com/kartoza/LEDET_BIMS/issues/304

peek 2018-11-07 13-00

### CODE REVIEW

**Comment by ann26** [comment]
There's a slight design problem issue:
When I logged in, the map control panel becomes very long like this:
screenshot from 2018-11-07 13-06-46

Thus, when legend is opened some of the button in the map control panel are hidden like this:
screenshot from 2018-11-07 13-08-51

In this case, the print map button is hidden behind the legend.

@carina1000 any ideas how to overcome this?

**Comment by ann26** [comment]
this PR might solve the design issue: https://github.com/kartoza/django-bims/pull/482

File: bims/static/js/views/olmap_layers.js [hunk 1]
```diff
@@ -200,7 +203,8 @@ define(['shared', 'backbone', 'underscore', 'jquery', 'ol', 'views/layer_style']
                             params: {
                                 layers: value.wms_layer_name,
                                 format: value.wms_format
-                            }
+                            },
+                            crossOrigin: 'anonymous'
```
Comment (dimasciput) [inline bims/static/js/views/olmap_layers.js]: @ann26 if you add this line then I couldn't access layers from maps.kartoza geoserver, here's the error message : 

image


@NyakudyaA any solution for this ?

Comment (NyakudyaA) [reply bims/static/js/views/olmap_layers.js]: @dimasciput let me see if I enable GeoServer CORS if it will work for these layers

Comment (NyakudyaA) [reply bims/static/js/views/olmap_layers.js]: @dimasciput cc @ann26 can you try to access GeoServer layers now and see if you can access them

Comment (dimasciput) [reply bims/static/js/views/olmap_layers.js]: @NyakudyaA I could access them, but if I show the layers from our GeoServer instances in testing and staging limpopobims, I will get the same error. What did you do on the GeoServer?  Maybe I can replicate the same process for our own instances.

Comment (dimasciput) [reply bims/static/js/views/olmap_layers.js]: @ann26 If I choose Topography or Aerial basemap layer I will get this message : 
image

And the basemap will not appear.

Comment (ann26) [reply bims/static/js/views/olmap_layers.js]: hi @ann26 in my local it does not work from the beginning, this code didn't change the way they behave in my local. I always got cors error either way

Comment (ann26) [reply bims/static/js/views/olmap_layers.js]: @dimasciput @NyakudyaA @gubuntu 
In my local, only biodiversity and Rivers NFEPA layers can be displayed (others I got 404) and the the basemap that is working is Topography, Openstreetmap, Aerial Photography and Bing (for terrain I got invalid key, for plain greyscale and B&W simply does not work I couldn't figure out why).

However due to cors error, the biodiversity layers can't be downloaded (although I remember it was working before) and the Topography and Aerial Photography basemap can't be downloaded either because of cors error (I tried to add `crossOrigin: 'anonymous'` but apparently it makes the map does not appear, so I removed the cross origin from them)

So the only layer and basemaps that I've tested and working to be downloaded are:
- Rivers NFEPA (layer)
- Openstreetmap (basemap)
- Bing (basemap)

Comment (gubuntu) [reply bims/static/js/views/olmap_layers.js]: afaik LEDET overlays come only from the local GeoNode and hence are on the same domain so there should be no CORS. 

All the basemaps come from third parties, surely their CORS behaviour should be consistent and able to be overcome? Same if we fetch WMS or WFS from third parties like maps.kartoza.com. 

Also @ann26 what do you mean by 'downloaded' in the context of pdf generation?

Comment (ann26) [reply bims/static/js/views/olmap_layers.js]: @gubuntu yes that's what I thought with the basemaps too, but apparently the ones that does not have cors problem are openstreetmap and bing.

I mean when we download the map, when the basemap or the layer has cors problem the result is an empty canvas as shown below:
map 58

while the maps that is successfully downloaded should be like this:
(basemap: bing, layer: Rivers NFEPA)
map 53

Comment (ann26) [reply bims/static/js/views/olmap_layers.js]: update: biodiversity layer has no cors problem

**Comment by ann26** [comment]
hi @gubuntu @dimasciput @NyakudyaA 
I added pop up that will notify user when they try to download basemap that raise CORS error like this:

peek 2018-11-15 14-59

Since the tainted canvas because of cors will remains tainted even though user select another basemap. So it needs to be reloaded.

Other basemap and layers that have no error will work normally (i.e. we can download the map with displayed layers and legend)

**Comment by ann26** [comment]
When there is no cors error:
Result in .pdf (format: A4 landscape)
peek 2018-11-15 16-41


When there is cors error:
peek 2018-11-15 16-42

**Comment by gubuntu** [comment]
expand what CORS means in your popup sice users won't have a clue what that is. 'there was a problem adding some third party layer(s) to the map' 

But we should really get around the CORS issues (in the next iteration) - i can't help there. please create another issue for that in the backlog

otherwise LGTM

**Comment by NyakudyaA** [comment]
@ann26 Maybe this might do the trick
https://gis.stackexchange.com/questions/71715/enabling-cors-in-openlayers/71726

**Comment by ann26** [comment]
Hi @gubuntu I have updated the wording in the pop up, as follows:
screenshot from 2018-11-16 07-34-35

thanks @NyakudyaA for the tips, I will add that in the new issue for CORS

**Comment by dimasciput** [comment]
I'm still not confident to put this on live server, it might break something, I just tested it in my local, and the layers are not showing up.

**Comment by timlinux** [comment]
My suggestion would be to implement a simple HTTP proxy on the BIMS server that will relay tile requests to the appropriate upstream server. That way all tile requests will appear to be against the same domain. For domains that already work we should not use the proxy as there will be a performance hit in taking this route.

See example of how we can simply build an http proxy:

https://stackoverflow.com/questions/4412581/seriously-simple-python-http-proxy

**Comment by lucernae** [comment]
> Maybe @timlinux @lucernae @dimasciput have a tehcnical solution to completely avoid CORS issues.

I don't know if this will helps, but I usually just proxied the resource to the django server that will serve them. So, for example when your map client request a tile server  in X, they don't request to A directly, but instead they request to your django server (the proxy, or any kind of accepted proxy). Then, that proxy will read which server to request (X url), do some http request to that URL and fetch the content. The content itself then passed on to that map client request.

In summary:

map client (your javascript code that generates pdf) --> proxy (django view or other proxy that can collect the tile url you need) --> the actual tile url (asking the content) --> proxy (accept the content from the tile url and repackage it to send to map client) --> your map client

Basically I do this when showing tiles/legends/thumbnails that is not in my domain. Because you can't control CORS policy of someone else's domain.
We also generated pdf report in server side whenever possible because you won't have to deal with CORS in that case.

**Comment by lucernae** [comment]
If you strictly have to use JavaScript to render the pages, I have used PhantomJS in the past for this kind of use case (printing tickets from HTML generated page).
But, PhantomJS is in suspended mode at the moment.
I don't know if this is relevant, but I will just share related article that describes these (server side javascript content generation).
http://phantomjs.org
https://aboullaite.me/phantomjs-generate-pixel-perfect-pdf-reports/
